pkg/server/ptypes/pipeline_ui.go

```diff
@@ -0,0 +1,479 @@
+
+	return rootNode, nil
+}
+
+// root returns the root ID and node.
+func (p uiPipelineProcessor) root() (string, *pb.UI_PipelineRunTreeNode, error) {
+	sorted := p.graph.KahnSort()
+
+	id, ok := sorted[0].(string)
+	if !ok {
+		return "", nil, fmt.Errorf("could not find root node")
+	}
+
+	node, ok := p.nodeIdx[id]
+	if !ok {
+		return "", nil, fmt.Errorf("could not find root node")
+	}
+
+	return id, node, nil
+}
+
+// processSubGraph performs transformations starting at the given inputId, and
+// collecting the results into the given outputNode.
+func (p uiPipelineProcessor) processSubGraph(inputId string, outputNode *pb.UI_PipelineRunTreeNode) error {
+	var nextIds []string
+	for _, v := range p.graph.OutEdges(inputId) {
+		nextIds = append(nextIds, v.(string))
+	}
+	sort.Strings(nextIds)
+	degree := len(nextIds)
+
+	if degree == 1 {
+		nextId := nextIds[0]
+		nextNode := p.nodeIdx[nextId]
+		inputPipelineId := p.jobIdx[inputId].Pipeline.GetPipelineId()
+		nextPipelineId := p.jobIdx[nextId].Pipeline.GetPipelineId()
+
+		if inputPipelineId != nextPipelineId {
+			invokeId, err := p.foldSubPipeline(nextId, inputPipelineId)
+			if err != nil {
+				return err
+			}
+			nextId = invokeId
+			nextNode = p.nodeIdx[nextId]
+		}
+
+		outputNode.Children.Nodes = append(outputNode.Children.Nodes, nextNode)
+
+		return p.processSubGraph(nextId, outputNode)
+	}
+
+	if degree > 1 {
+		virtualId, virtualNode, err := p.foldBranches(inputId, outputNode)
+		if err != nil {
+			return err
+		}
+
+		// Continue processing the parent graph from the new
+		// virtual node, but accumulating into the output node.
+		if err := p.processSubGraph(virtualId, outputNode); err != nil {
+			return err
+		}
+
+		// Process all the input children as sub-graphs,
+		// accumulating into the virtual node.
+		for _, nextId := range nextIds {
+			nextNode := p.nodeIdx[nextId]
+			inputJob := p.jobIdx[inputId]
+			nextJob := p.jobIdx[nextId]
+			inputPipelineId := inputJob.Pipeline.GetPipelineId()
+			nextPipelineId := nextJob.Pipeline.GetPipelineId()
+
+			if inputPipelineId != nextPipelineId {
+				invokeId, err := p.foldSubPipeline(nextId, inputPipelineId)
+				if err != nil {
+					return err
+				}
+				nextId = invokeId
+				nextNode = p.nodeIdx[nextId]
+			}
+
+			virtualNode.Children.Nodes = append(virtualNode.Children.Nodes, nextNode)
+
+			if err := p.processSubGraph(nextId, nextNode); err != nil {
+				return err
+			}
+		}
+
+		p.inferNodeAttrsFromChildren(virtualNode)
+	}
+
+	return nil
+}
+
+// foldBranches replaces parallel branches of a graph with a single node.
+//
+// Before:
+//
+//	  A
+//	┌─┴─┐
+//	B   D
+//	C   E
+//	└─┬─┘
+//	  F
+//
+// After:
+//
+//	A
+//	V
+//	F
+//
+// It returns the new node and assumes the parent function will perform further
+// processing to embed the original children within the new node.
+func (p uiPipelineProcessor) foldBranches(inputId string, outputNode *pb.UI_PipelineRunTreeNode) (string, *pb.UI_PipelineRunTreeNode, error) {
```

Comment:
Just a small style nit, since this func is kinda long, we typically format it to be more condensed. Something like:

```suggestion
func (p uiPipelineProcessor) foldBranches(
	inputId string,
	outputNode *pb.UI_PipelineRunTreeNode,
) (string, *pb.UI_PipelineRunTreeNode, error) {
```